ops/model.py

```diff
@@ -797,6 +961,313 @@ def __init__(self, name: str, address_info: '_AddressDict'):
+                 am_owner: bool = False):
+        if not label and not id:
+            raise RuntimeError('Provide at least one of (`label`|`id`).')
+        if am_owner and revision is None:
+            # This should only happen in testing, when one instantiates Secret manually.
+            logger.error('secret {!r} (owned by this unit) has no revision!'
+                         ' Things might malfunction'.format(id))
+
+        self._backend = backend
+
+        if isinstance(id, str):
+            if not id.startswith('secret:'):  # todo: regex validation?
+                raise InvalidSecretIDError(
+                    'secret secret ID {!r} should start with `secret:`'.format(id)
+                )
+            if id.strip() != id:
+                # FIXME one of the hook tools returns some whitespace around
+                #  the secret-id. We should catch that earlier and trim it there.
+                logger.warning(
+                    'invalid characters in secret ID '
+                    '{!r}: something wrong'.format(id)
+                )
+                id = id.strip()
+
+        self._id = id
+        self._label = label
+        self._revision = revision
+        self._am_owner = am_owner
+
+    @property
+    def id(self) -> Optional[str]:
+        """Unique identifier for this secret.
+
+        The secret ID will be known (not None) only if you either:
+            - obtain this secret by `self.[unit|app].add_secret(...)`
+            - obtain this secret by `self.model.get_secret(id=<some-id>)`
+            - obtain this secret from a SecretEvent instance.
+
+        If you obtained this secret by `self.model.get_secret(label=<some-label>)`,
+            then the secret ID will be unknown.
+            The secret will then be locally addressable only by its label.
+        """
+        return self._id
+
+    @property
+    def label(self) -> Optional[str]:
+        """Label by which this secret is locally known.
+
+        The label will be not None only if you either:
+            - obtain this secret by `self.[unit|app].add_secret(label=<some-label>, ...)`
+            - obtain this secret by `self.model.get_secret(label=<some-label>)`
+            - obtain this secret from a SecretEvent instance.
+
+        If you obtained this secret by `self.model.get_secret(id=<some-id>)`,
+            then the label for this secret will be unknown.
+            The secret will then be locally addressable only by its ID.
+
+        This label is locally unique.
+        """
+        return self._label
+
+    @property
+    def revision(self) -> int:
+        """The revision number for this secret."""
+        if self._revision is None:
+            if self._am_owner:
+                # use _revision if you know what you're doing (testing, presumably)
+                raise RuntimeError("This secret has no revision.")
+            else:
+                raise RuntimeError("Only secret owners can see the revision number.")
```

Comment:
Do we actually want a `RuntimeError` here instead of raising a more specific exception which can be caught for flow control?

Reply:
Yes, I meant to do that but I forgot. Done!